crates/proof-of-sql/src/sql/proof_exprs/range_check.rs

```diff
@@ -23,44 +116,315 @@ fn decompose_scalar_to_words<'a, S: Scalar + 'a>(
+        vec![
+            (
+                S::one(),
+                vec![
+                    Box::new(inverted_word_values_plus_alpha as &[_]),
+                    Box::new(word_values_plus_alpha as &[_]),
+                ],
+            ),
+            (-S::one(), vec![Box::new(input_ones as &[_])]),
+        ],
+    );
+}
+
+/// Argue that the sum of all words in each row, minus the count of each word
+/// multiplied by the inverted word value, is zero.
+///
+/// ```text
+/// ∑ (I₀ + I₁ + I₂ + I₃ - (C * IN)) = 0
+/// ```
+///
+/// Where:
+/// - `I₀, I₁, I₂, I₃` are the inverted word columns.
+/// - `C` is the count of each word.
+/// - `IN` is the inverted word values column.
+#[allow(clippy::missing_panics_doc)]
+fn prove_row_zero_sum<'a, S: Scalar + 'a>(
+    builder: &mut FinalRoundBuilder<'a, S>,
+    word_counts: &'a mut [i64],
+    alloc: &'a Bump,
+    scalars: &[S],
+    inverted_word_columns: &[&mut [S]],
+    alpha: S,
+) {
+    // Produce an MLE over the counts of each word value
+    builder.produce_intermediate_mle(word_counts as &[_]);
+
+    // Allocate row_sums from the bump allocator, ensuring it lives as long as 'a
+    let row_sums = alloc.alloc_slice_fill_with(scalars.len(), |_| S::ZERO);
+
+    // Iterate over each column and sum up the corresponding row values
+    for column in inverted_word_columns {
+        // Iterate over each scalar in the column
+        for (i, inv_word) in column.iter().enumerate() {
+            row_sums[i] += *inv_word;
+        }
+    }
 
-        // Invert all the terms in the column at once
-        slice_ops::batch_inversion(&mut terms_to_invert);
+    // Pass the row_sums reference with the correct lifetime to the builder
+    builder.produce_intermediate_mle(row_sums as &[_]);
+
+    // Allocate and store the row sums in a Box using the bump allocator
+    let row_sums_box: Box<_> =
+        Box::new(alloc.alloc_slice_copy(row_sums) as &[_]) as Box<dyn MultilinearExtension<S>>;
+
+    let inverted_word_values_plus_alpha: &mut [S] = alloc.alloc_slice_fill_with(256, |i| {
+        S::try_from(i.into()).expect("word value will always fit into S") + alpha
+    });
+
+    slice_ops::batch_inversion(&mut inverted_word_values_plus_alpha[..]);
+
+    // Now pass the vector to the builder
+    builder.produce_sumcheck_subpolynomial(
+        SumcheckSubpolynomialType::ZeroSum,
+        vec![
+            (S::one(), vec![row_sums_box]),
+            (
+                -S::one(),
+                vec![
+                    Box::new(word_counts as &[_]),
+                    Box::new(inverted_word_values_plus_alpha as &[_]),
+                ],
+            ),
+        ],
+    );
+}
 
-        // Assign the inverted values back to the inverted_word_columns
-        inverted_word_columns[i].copy_from_slice(&terms_to_invert);
+/// Verify that the prover claim is correct.
+#[allow(clippy::missing_panics_doc)]
+pub fn verifier_evaluate_range_check<C>(
```

Comment:
There needs to be a check that
```
sum w_eval[i] * 2^i == scalars_eval
```

Reply:
So far I have this:

```rust
#[derive(Debug, Serialize)]
pub struct RangeCheckTestPlan {
    pub column: ColumnRef,
}

impl ProofPlan for RangeCheckTestPlan {
    #[doc = " Form components needed to verify and proof store into `VerificationBuilder`"]
    fn verifier_evaluate<S: Scalar>(
        &self,
        builder: &mut VerificationBuilder<S>,
        accessor: &IndexMap<ColumnRef, S>,
        _result: Option<&OwnedTable<S>>,
        one_eval_map: &IndexMap<TableRef, S>,
    ) -> Result<TableEvaluation<S>, ProofError> {
        let scalar_eval = accessor[&self.column];
        verifier_evaluate_range_check(builder, one_eval_map, scalar_eval);

        Ok(TableEvaluation::new(
            vec![accessor[&self.column]],
            one_eval_map[&self.column.table_ref()],
        ))
    }
```

So we have ```scalar_eval``` set to this: 

``` let scalar_eval = accessor[&self.column];``` 

What is ```sum w_eval[i]```?

Is it derived this way?

```rust
/// Verify that the prover claim is correct.
#[allow(clippy::missing_panics_doc)]
pub fn verifier_evaluate_range_check<C>(
    builder: &mut VerificationBuilder<'_, C>,
    one_eval_map: &IndexMap<TableRef, C>,
    scalar_eval: C,
) where
    C: Scalar,
{
    let alpha = builder.consume_post_result_challenge();
    let mut w_plus_alpha_inv_evals: Vec<_> = Vec::with_capacity(31);

    let mut sum = C::ZERO;

    // Consume the (wᵢⱼ)  and (wᵢⱼ + α)⁻¹ MLEs
    for i in 0..31 {
        let one_eval = one_eval_map
            .values()
            .next()
            .expect("one_eval_map should have at least one value");

        let w_eval = builder.consume_mle_evaluation();
        let words_plus_alpha_inv = builder.consume_mle_evaluation();

        // Store the evaluations of (wᵢⱼ + α)⁻¹
        w_plus_alpha_inv_evals.push(words_plus_alpha_inv);

        // Verify that:
        // (wᵢⱼ + α)⁻¹ * wᵢⱼ + α * (wᵢⱼ + α)⁻¹ - 1 = 0
        let word_eval = words_plus_alpha_inv * w_eval + (alpha * words_plus_alpha_inv);
        sum += word_eval * C::from((2_i32).pow(i));

        builder.produce_sumcheck_subpolynomial_evaluation(
            SumcheckSubpolynomialType::Identity,
            word_eval - *one_eval,
        );
    }

    assert_eq!(sum, scalar_eval);
```